### PR TITLE
Issue/1101 - Delete stale systems when syncing child garden

### LINKED ISSUE #1101: Garden Sync with incompatible systems
It looks like the following scenario results in a complete failure of the sync:

1. Parent and child are synced
2. Stop the parent garden so it doesn't receive updates from the child garden
2. A system is removed on the child
3. A "new" system with the same name+version, ~but different commands~, is created on the child. EDIT: The commands don't need to be different - just removing and re-adding the system (so it has a different ID)
4. Start the parent
4. Parent initiates a sync

The parent will attempt to update the "replaced" system definition but will fail because of the commands mismatch.

### PR DESCRIPTION
**Pull request by jlrpnbbngtn**
Closes #1101

To verify the fix for the corner case (running BG in a parent-child configuration,, taking the parent offline, changing Systems on the child, bringing the parent back up and then syncing), a straightforward way is to follow the steps in #1101 while tailing the log of the parent and then seeing that an error is no longer thrown. One could also track the entries in the database across the operations to make sure they are consistent with the actions taken and expected results. This was done with single parent and single child with each of http and stomp entrypoints and tested with multiple children and a grandchild with http. Corner case error appears to be resolved in all of these scenarios.

For the general case, two unit tests are provided to verify that any unknown Systems associated with a remote garden are removed prior to saving the garden to the database.

### CODE REVIEW

File: src/app/test/db/mongo/models_test.py [hunk 1]
```diff
@@ -612,3 +645,35 @@ def test_only_one_local_garden_may_exist(self, testgarden):
...
+        new_vers = set(
+            map(
+                lambda x: str(getattr(x, "version")), child_garden.systems  # noqa: B009
+            )
+        )
+
+        assert self.v1_str not in new_vers and self.v2_str in new_vers
+        assert new_ids.intersection(orig_ids) == set()
+
+        child_system_v2.delete()
```
Comment (scott-taubman) [inline src/app/test/db/mongo/models_test.py:679]: This cleanup seems like it should happen in the fixture.

File: src/app/test/db/mongo/models_test.py [hunk 2]
```diff
@@ -612,3 +645,35 @@ def test_only_one_local_garden_may_exist(self, testgarden):
         should raise an exception"""
         with pytest.raises(NotUniqueError):
             Garden(name=f"not{testgarden.name}", connection_type="LOCAL").save()
+
+    def test_child_garden_system_update(self, child_garden, child_system_v2):
```
Comment (scott-taubman) [inline src/app/test/db/mongo/models_test.py:649]: This test is a good one to have, but should we also have one that tests that the update works when a new system with a matching "triple" shows up?  That's the case that inspired these changes in the first place.

File: src/app/beer_garden/db/mongo/models.py [hunk 3]
```diff
@@ -755,8 +757,60 @@ class Garden(MongoModel, Document):
...
 
     def deep_save(self):
-        for system in self.systems:
-            system.save()
+        logger = logging.getLogger(self.__class__.__name__)
+
+        # if the call to this method is on a child garden object, we ensure that
+        # when saving the systems, unknowns are deleted
+        if self.connection_type != "LOCAL":
+            # import moved here to avoid a circular import loop
```
Comment (scott-taubman) [inline src/app/beer_garden/db/mongo/models.py:765]: It feels appropriate to me to have this logic contained in its own function.  Then the `deep_save` would just look something like:

```python
def deep_save(self)
    if self.connection_type != "LOCAL":
        self._update_associated_systems()
        
    self.save()
```

**Review by scott-taubman — CHANGES_REQUESTED** [review]

File: src/app/beer_garden/db/mongo/models.py (hunk 3 above)
Comment (jlrpnbbngtn) [reply src/app/beer_garden/db/mongo/models.py:765]: This is a pattern I see in your code a lot. I'm not against the habit, so sure. Done.

**Review by scott-taubman — APPROVED** [review]
I think we can merge this, but let's wait until we fix the build issue on develop first (hopefully #1138 does the trick) before introducing additional changes.
